### ISSUE TITLE
Rust: Permissions Question

### ISSUE THREAD

**Issue by CodeMan99** [devcontainers/images#851]
## Scenario

The rust image currently has `umask 022` set.

```shell
$ docker run -it --rm --user root mcr.microsoft.com/devcontainers/rust:1-1-bullseye
root ➜ / $ umask
0022
root ➜ / $ grep ^rustlang /etc/group
rustlang:x:999:vscode
```

Should a build step, such as a _feature_ use `cargo install` the resulting permissions prevent group write usage.

```shell
root ➜ / $ cargo install --quiet bacon
root ➜ / $ ls -ld $CARGO_HOME/registry
drwxr-sr-x 5 root rustlang 4096 Nov 14 19:04 /usr/local/cargo/registry
```

Which then in turn breaks user-level usage of other cargo commands.

```shell
root ➜ / $ su vscode
vscode ➜ / $ cd
vscode ➜ ~ $ cargo install cargo-watch
    Updating crates.io index
  Downloaded cargo-watch v8.4.1
error: failed to download replaced source registry `crates-io`

Caused by:
  failed to open `/usr/local/cargo/registry/cache/index.crates.io-6f17d22bba15001f/cargo-watch-8.4.1.crate`

Caused by:
  Permission denied (os error 13)
```

## Question

What is the correct course of action?

1. Set `umask 002` here, at the image level? Perhaps in `/etc/profile`?
2. Set `umask 002` inside feature `install.sh` scripts?

**Comment by samruddhikhandale** [comment]
Hi 👋 

Thanks for opening a detailed issue.

> What is the correct course of action?

As `cargo` is installed by the [Rust Feature](https://github.com/devcontainers/features/tree/main/src/rust), it makes sense to update permissions in the Feature. However, looking at https://github.com/devcontainers/features/blob/main/src/rust/install.sh#L156, I do see `umask 0002` set. However, I think there's a bug in this line https://github.com/devcontainers/features/blob/main/src/rust/install.sh#L163, it updates permissions only for the specific directory, instead of doing that recursively.

I believe, updating `chmod g+r+w+s "${RUSTUP_HOME}" "${CARGO_HOME}"` to 👇  should help.

```
chmod -R g+r+w+s "${RUSTUP_HOME}" "${CARGO_HOME}"
```

@CodeMan99 Let me know if that makes sense to you, thanks!

**Comment by samruddhikhandale** [comment]
Opened https://github.com/devcontainers/features/pull/754

**Comment by CodeMan99** [comment]
> Let me know if that makes sense to you, thanks!

I am not completely sure to be honest. I don't think the feature creates `$CARGO_HOME/regitstry`. I believe that folder is only created after usage of `cargo install`.

So the _rust feature_ still isn't responsible for the creation of the problem directory.

**Comment by samruddhikhandale** [comment]
I think I missed the part where you switch the users between running `cargo install command`. As the `registry` is created by the `root` user, it won't automatically allow the `vscode` user to use it unless the permissions are updated. 

Curious, @CodeMan99 why do you switch users? I believe these Features/images were created hoping that the same user is been used for lifecycle of the dev container. Else, it would cause permission issues as seen in this issue ^

**Comment by CodeMan99** [comment]
> Curious, @CodeMan99 why do you switch users? I believe these Features/images were created hoping that the same user is been used for lifecycle of the dev container. Else, it would cause permission issues as seen in this issue ^

Usage of root is build-like usage. For example, using `cargo install` inside a feature `install.sh`. My understanding that executing that script will always be `root`. Is that an incorrect assumption?

Then, during development `cargo` will need write access to that folder. Including `cargo build`. The `$CARGO_HOME/registry` folder is created with the correct group (rustlang) in my example.

**Comment by samruddhikhandale** [comment]
> My understanding that executing that script will always be root. Is that an incorrect assumption?

Yes, that is correct. All Feature's install scripts are run with the `root` user.

With the `Rust` Feature, `CARGO_HOME` directory is correctly owned by the `vscode` user. However, as you are running `cargo` commands with the `root` user, it creates the `registry` folder with `root` permissions. The non-root user cannot modify the files that were created by `root` as basically you can't mix something like `sudo cargo install` and `cargo install`.

Hence, I believe you should do either of the following to fix the issue - 
- Always stick with a single user `vscode` or `root`, else, you will run into this issue
    - Else, manually give permissions for the `registry` directory to the `vscode` user
- When running with `root`, make sure to run commands as `vscode` user (eg. vscode -c "cargo install bacon")

**Comment by CodeMan99** [comment]
> However, as you are running `cargo` commands with the `root` user, it creates the `registry` folder with `root` permissions.

The folder is created with exactly these permissions.

```shell
root ➜ / $ cargo install --quiet bacon
root ➜ / $ ls -ld $CARGO_HOME/registry
drwxr-sr-x 5 root rustlang 4096 Nov 14 19:04 /usr/local/cargo/registry
```

The `vscode` user is added to the `rustlang` group by the [rust feature](https://github.com/devcontainers/features/blob/d53b9d1816bac7a62f9ba5f20723eef428042d1f/src/rust/install.sh#L160). The only thing that needs to change is the write permissions for the group.

> * Always stick with a single user `vscode` or `root`, else, you will run into this issue

This advice does not resolve the issue. Example devcontainer.json file:

```jsonc
{
     "image": "mcr.microsoft.com/devcontainers/rust:1-1-bullseye",
     "features": {
         "ghcr.io/lee-orr/rusty-dev-containers/bacon:0": {}
     }
}
```

This very simple configuration creates the problem:

1. `cargo install bacon` is run in the install.sh of the feature.
2. The `remoteUser` is `vscode`.

---

So again, what is the correct course of action?

Use of `umask 002` in the image's `/etc/profile` file? Or use of `umask 002` in the install.sh script of any given feature?

**Comment by CodeMan99** [comment]
Let me clarify a little - "the course of action" is a question of _who_ needs to own this knowledge.

1. Setting the `umask 002` in `/etc/profile` would be an image-wide change, affecting all system user's default. Only image maintainers would be **required** to have knowledge.
2. Setting the `umask 002` in any given feature is a very specific change that requires documentation. Both documentation maintainers and community feature maintainers are **required** to have knowledge.
3. Just providing a guideline puts the ownership of this issue in the hands of any user that wants to use `cargo install` as a build step (Dockerfile or using a feature).

**Comment by gamife** [comment]
> Opened [devcontainers/features#754](https://github.com/devcontainers/features/pull/754)

This PR is what I want.
Now I have an error using the `cargo build` command due to permission being denied. Perhaps it's because I used cargo to install some tools in the dockerfile before run rust feature script, so the gid in the registry directory is `root`
```bash
non-root@702362714ab7:~$ ls -l $CARGO_HOME
total 24
drwxrwxrwx 1 root root     4096 May 22 08:04 bin
-rwxr-xr-x 1 root rustlang  690 May 21 16:06 config.toml
-rw-rw-rw- 1 root root      308 May 14 03:57 env
drwxrwxr-x 1 root root     4096 May 22 08:04 registry

non-root@702362714ab7:~$ ls -l $RUSTUP_HOME
total 20
drwxrwxrwx 1 root root 4096 May 22 09:04 downloads
-rw-rw-rw- 1 root root  151 May 14 03:58 settings.toml
drwxrwxrwx 1 root root 4096 May 22 09:04 tmp
drwxrwxrwx 1 root root 4096 May 22 09:04 toolchains
drwxrwxrwx 1 root root 4096 May 22 09:04 update-hashes
```
Can this PR be merged?
